core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java

```diff
@@ -76,33 +89,36 @@ public static boolean inodeUfsIsSynced(Inode inode, String ufsFingerprint) {
           && inodeFile.getUfsFingerprint().equals(ufsFingerprint)
           && !inodeFile.getUfsFingerprint().equals(Constants.INVALID_UFS_FINGERPRINT);
     } else {
-      // ufs fingerprint must exist.
-      // TODO(gpang): Currently, directory fingerprints are not considered, because directories are
-      // created/modified more frequently than files.
-      isSyncedPersisted =
-          inode.isPersisted() && !Constants.INVALID_UFS_FINGERPRINT.equals(ufsFingerprint);
+      isSyncedPersisted = inode.isPersisted() && inode.getUfsFingerprint().equals(ufsFingerprint);
     }
     return isSyncedPersisted || isSyncedUnpersisted;
   }
 
   /**
    * A class describing how to sync an inode with the ufs.
    * A sync plan has several steps:
-   * 1. delete: the inode should be deleted
-   * 2. syncChildren: the inode is a directory, and the children should be synced
-   * 3. loadMetadata: the inode metadata should be loaded from UFS
+   * 1. updateDirectory: the directory inode should update permissions from UFS directory
+   * 2. delete: the inode should be deleted
+   * 3. syncChildren: the inode is a directory, and the children should be synced
+   * 4. loadMetadata: the inode metadata should be loaded from UFS
    */
   public static final class SyncPlan {
+    private boolean mUpdateDirectory;
```

Comment:
Could this be extended to files as well, where only the metadata needs to be repaired?

Reply:
Currently, we cannot do that, since fingerprint includes both content and metadata. However, I left a TODO in Fingerprint to partition metadata from content, which would enable metadata-only updates.